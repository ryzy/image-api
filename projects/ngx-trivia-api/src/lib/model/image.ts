export interface ImageLinks {
  /**
   * HTML link to the image (browse'able from browser)
   */
  html: string;

  large: string;
  regular: string;
  small: string;
  thumb: string;
}

/**
 * Object present when license is different than PublicDomain
 */
export interface ImageAuthor {
  name?: string;
  url?: string;
  avatar?: string;
  twitter?: string;
}

export interface Image {
  id: string;
  caption: string;
  source: 'Google' | 'Unsplash';
  license: 'cc_publicdomain' | 'cc_attribute';
  location?: string;
  links: ImageLinks;
  author?: ImageAuthor;
}
