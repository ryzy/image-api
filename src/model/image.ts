export interface ImageLinks {
  html: string;
  full: string;
  raw: string;
  regular: string;
  small: string;
  thumb: string;
}
export interface ImageAuthor {
  source: 'Unsplash';
  name: string;
  url: string;
  avatar: string;
  twitter: string;
}

export interface Image {
  id: string;
  caption: string;
  location: string;
  links: ImageLinks;
  author: ImageAuthor;
}

export function unsplashToImage(data: any = {}): Image {
  return {
    id: data.id,
    caption: data.description,
    location: data.location && data.location.title,
    links: {
      html: data.links && data.links.html,
      ...data.urls,
    },
    author: {
      source: 'Unsplash',
      name: data.user && data.user.name,
      url: data.user && data.user.links && data.user.links.html,
      avatar:
        data.user && data.user.profile_image && data.user.profile_image.large,
      twitter: data.user && data.user.twitter_username,
    },
  };
}
