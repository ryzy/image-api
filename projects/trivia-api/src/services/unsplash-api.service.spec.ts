import { Test, TestingModule } from '@nestjs/testing';
import { HttpModule, HttpService } from '@nestjs/common';
import { marbles } from 'rxjs-marbles';
import { of, throwError } from 'rxjs';
import { AxiosError, AxiosResponse } from 'axios';

import { UnsplashApiService, unsplashToImage } from './unsplash-api.service';
import { mockImage1 } from '../../../../test/fixtures/image';
import { mockUnsplashImage } from '../../../../test/fixtures/unsplash-image';

const mockHttpError: AxiosError = {
  name: 'Error',
  message: 'Some Error',
  config: {},
  response: {
    data: {
      errors: ['Some error', 'Some another error'],
    },
  } as AxiosResponse,
};

describe('UnsplashApiService', () => {
  let app: TestingModule;
  let http: HttpService;
  let service: UnsplashApiService;

  beforeAll(async () => {
    app = await Test.createTestingModule({
      imports: [HttpModule],
      providers: [UnsplashApiService],
    }).compile();

    http = app.get<HttpService>(HttpService);
    service = app.get<UnsplashApiService>(UnsplashApiService);
  });

  test('#unsplashToImage', () => {
    expect(unsplashToImage()).toMatchSnapshot();
    expect(unsplashToImage(mockUnsplashImage)).toEqual(mockImage1);
  });

  describe('getImage', () => {
    it(
      'should get image',
      marbles((m) => {
        jest.spyOn(http, 'get').mockReturnValue(of({ data: mockUnsplashImage }));
        m.expect(service.getImage()).toBeObservable('(v|)', { v: mockImage1 });
      }),
    );

    it(
      'should catch error',
      marbles((m) => {
        jest.spyOn(http, 'get').mockReturnValue(throwError(mockHttpError));
        m.expect(service.getImage()).toBeObservable('#', undefined, {
          name: 'Error',
          message: {
            errors: mockHttpError.response!.data.errors,
            status: 500,
          },
        });
      }),
    );
  });

  describe('getImages', () => {
    it(
      'should get images when no query',
      marbles((m) => {
        jest.spyOn(http, 'get').mockReturnValue(of({ data: [mockUnsplashImage] }));
        m.expect(service.getImages()).toBeObservable('(v|)', {
          v: [mockImage1],
        });
      }),
    );

    it(
      'should get images',
      marbles((m) => {
        jest.spyOn(http, 'get').mockReturnValue(of({ data: { results: [mockUnsplashImage] } }));
        m.expect(service.getImages('some query')).toBeObservable('(v|)', {
          v: [mockImage1],
        });
      }),
    );
  });
});
