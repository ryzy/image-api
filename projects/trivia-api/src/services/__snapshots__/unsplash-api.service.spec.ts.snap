// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnsplashApiService #unsplashToImage 1`] = `
Object {
  "author": Object {
    "avatar": undefined,
    "name": undefined,
    "twitter": undefined,
    "url": undefined,
  },
  "caption": undefined,
  "id": undefined,
  "license": "cc_attribute",
  "links": Object {
    "html": undefined,
    "large": undefined,
    "regular": undefined,
    "small": undefined,
    "thumb": undefined,
  },
  "location": undefined,
  "source": "Unsplash",
}
`;
